### ISSUE TITLE
Defining Costum Hamiltonian

### ISSUE THREAD

**Issue by thorben-frank** [netket/netket#293]
First of all: Great project!! 
I have a question about the custom Hamiltonians:
When I run your pre-implemented Heisenberg Hamiltonian after iteration 0 my proposed energy is already somewhere below zero same holds for the pre-implemented TFI model. 
However, when I am building these two Hamiltonians via the LocalOperator class I get the following behavior:

- For the TFI model the energy after iteration 0 is the same for both the custom build Hamiltonian and the pre-implemented version.

- For the Heisenberg model it differs in such a way that for the custom build Hamiltonian the energy after iteration zero is way above zero. 

So I was wondering where this behavior comes from?

**Comment by femtobit** [comment]
Hi. Some more information on your setup might help us figure out what is happening. Could you provide the code you have used to specify your custom Hamiltonians? Also, which algorithm and parameters are you using for the energy minimization? Which machine are you using (and how are you initializing its weights)?

**Comment by thorben-frank** [comment]
Hey, 
at the end my code for creating and testing my custom H's. 
For the energy minimization I am using the SR with SGD as optimizer with the following parameters:

- diagonal shift for the S matrix = 0.01
- learning rate SGD = 0.01
- in the SR I am using the iterative solver

As machine I am using the RBM Spin with:

- randomly initialized parameters (seed = 123, sigma = 0.01)

```cpp
#include "netket.hpp"
#include <iostream>
#include <Eigen/Dense>
#include <Eigen/KroneckerProduct>

std::vector<std::vector<std::complex<double>>> matrixToList(Eigen::MatrixXcd A){

  std::vector<std::vector<std::complex<double>>> b(A.rows());

  std::vector<std::complex<double>> ARow;
  ARow.resize(A.cols());

  for(int i = 0; i < A.rows(); i++){
    Eigen::Map<Eigen::VectorXcd>( &ARow[0], A.cols()) = A.row(i);
    b[i] = ARow;
  }
  
  return b;
}

void constructTFI(const int length = 12, const double J = 1., const double h = 0.5) {
  
  // construct graph
  std::array<int, 2> edge;
  std::vector<std::array<int,2>> edges(length);
  for(int i = 0; i<length; i++){
    if(i < (i + 1) % length){
      edge[0] = i;
      edge[1] = (i + 1) % length;
    }
    else{
      edge[0] = (i + 1) % length;
      edge[1] = i;
    }
    edges[i] = edge;
  }
  
  netket::CustomGraph g = netket::CustomGraph(edges);
  
  // construct Hilbert space from graph
  double spin = 0.5;
  netket::Spin hi = netket::Spin(g, spin);

  // construct Hamiltonian
  Eigen::MatrixXcd sx,sy,sz;
  sx.resize(2,2); sy.resize(2,2); sz.resize(2,2);
  sx << 0. + 0i, 1. + 0i, 1. + 0i, 0. + 0i;
  sz << 1. + 0i, 0. + 0i, 0. + 0i, -1. + 0i;
  
  Eigen::MatrixXcd Op1 = - h * sx;
  
  std::vector<std::vector<std::vector<std::complex<double>>>> operators(length * 2);
  std::vector<int> site1(1);
  std::vector<std::vector<int>> sites(length*2);
  int i = 0;
  for(i; i < length; i++){ 
    operators[i] = matrixToList(Op1);
    site1[0] = i;
    sites[i] = site1;
  }

  std::vector<int> site2(2);
  Eigen::MatrixXcd Op2 = - J * Eigen::kroneckerProduct(sz,sz);
  for(i; i < length*2; i++){
    operators[i] = matrixToList(Op2);
    site2[0] = i-length;
    site2[1] = (i-length + 1) % length;
    sites[i] = site2;
  }

  // Costum and pre implemented Hamiltonian
  netket::LocalOperator ham = netket::LocalOperator(hi, operators, sites);
  netket::Ising hamIsing = netket::Ising(hi, h, J);

  // checking that hamiltonian gives expected energy
  netket::eddetail::result_t results = netket::lanczos_ed(ham);
  std::cout << "Result costum Ising Hamiltonian: " << results.eigenvalues()[0] << std::endl;
  results = netket::lanczos_ed(hamIsing);
  std::cout << "Result pre implemented Ising Hamiltonian: " << results.eigenvalues()[0] << std::endl;
}

void constructHeisenberg(int length = 12, double Jx = 1, double Jy = 1, double Jz = 1){

  // construct graph
  std::array<int, 2> edge;
  std::vector<std::array<int,2>> edges(length);
  for(int i = 0; i<length; i++){
    if(i < (i + 1) % length){
      edge[0] = i;
      edge[1] = (i + 1) % length;
    }
    else{
      edge[0] = (i + 1) % length;
      edge[1] = i;
    }
    edges[i] = edge;
  }
  //  std::cout << edges << std::endl;
  netket::CustomGraph g = netket::CustomGraph(edges);
  
  // construct Hilbert space from graph
  double spin = 0.5;
  double totalSz = 0;
  netket::Spin hi = netket::Spin(g, spin, totalSz);

  // construct Hamiltonian
  Eigen::MatrixXcd sx,sy,sz;
  sx.resize(2,2); sy.resize(2,2); sz.resize(2,2);
  sx << 0. + 0i, 1. + 0i, 1. + 0i, 0. + 0i;
  sy << 0. + 0i, 0. - 1i, 0. + 1i, 0. + 0i;
  sz << 1. + 0i, 0. + 0i, 0. + 0i, -1. + 0i;
  
  Eigen::MatrixXcd Op = ( Jx * Eigen::kroneckerProduct(sx,sx) + Jy * Eigen::kroneckerProduct(sy,sy) + Jz * Eigen::kroneckerProduct(sz,sz) );
  
  std::vector<std::vector<std::vector<std::complex<double>>>> operators(length);
  std::vector<int> site(2);
  std::vector<std::vector<int>> sites(length);
  
  for(int i = 0; i < length; i++){
    operators[i] = matrixToList(Op);
    site[0] = i;
    site[1] = (i + 1) % length;
    sites[i] = site;
  }

  // Costum and pre implemented Hamiltonian
  netket::LocalOperator ham = netket::LocalOperator(hi, operators, sites);
  netket::Heisenberg hamHeis = netket::Heisenberg(hi);

  // checking that hamiltonian gives expected energy
  netket::eddetail::result_t results = netket::lanczos_ed(ham);
  std::cout << "Result costum Heisenberg Hamiltonian: " << results.eigenvalues()[0] << std::endl;
  results = netket::lanczos_ed(hamHeis);
  std::cout << "Result pre implemented Heisenberg Hamiltonian: " << results.eigenvalues()[0] << std::endl;
}

int main() {
  MPI_Init(NULL, NULL);
  constructTFI();
  constructHeisenberg();
  MPI_Finalize();
}
```

**Comment by kchoo1118** [comment]
This difference is because in the pre-implemented Heisenberg Hamiltonian, the Marshall sign rule is used. Perhaps in the custom Hamiltonian you built, you didn't use the sign rule.

**Comment by thorben-frank** [comment]
alright. Thanks for the quick responses!

**Comment by gcarleo** [comment]
That's most likely the reason, yes! Also notice that the C++ APIs are
typically not meant for direct use (I. E. They might have large changes
even between point releases, whereas the python APIs are stable). In this
case the python APIs are definitely the recommended way to set up a custom
hamiltonian!

**Comment by chaeyeunpark** [comment]
Hi all. I came here as I noticed this issue from Thorben.

So what we found is that the default python implementation of the Heisenberg model uses one of the stoquastic versions of the Hamiltonian. So instead of [XX+YY+ZZ], it uses [-(XX+YY)+ZZ] when the graph is bipartite. This Hamiltonian can be obtained from the original XXX model by applying Z gates to spins in one of the sublattices. Of course, the ground state energy does not change as this is local basis transformation but some caution is necessary if one wants to use the ground state wavefunction. 

We were a bit confused at first as the Marshall's sign rule is more about the wave functions rather than the Hamiltonian. So I am adding our interpretation based on stoquasticitiy to help some people who encounter the same problem.
